### PR TITLE
LibGfx/PNG: The promised PR

### PR DESCRIPTION
**Pull request by LucasChollet**
Clearly a good citizen PR as it has:

- More tests
- More AK::Stream usage
- Less :thousandcatdogstare: memory management

cc @nico

-------

**Tests: Move png test image into png/ subfolder**

**Tests: Add a test for animated PNG**

I might have broken APNG support at some point, so let's add a test to
prevent someone to break it on master.

The image is in the public domain:
https://en.wikipedia.org/wiki/File:Animated_PNG_example_bouncing_beach_ball.png

**LibGfx/PNG: Don't use a full context to sniff bytes**

No need to create a full context to just compare a few bytes. This is
mainly needed because the PNG context will soon require the creation of
a Stream.

**LibGfx/PNG: Factorize the operation to read a filter**

**AK: Rename the `const` overload of `FixedMemoryStream::bytes()`**

Due to overload resolutions rules, this simple code provokes a crash:

ReadonlyBytes readonly_bytes{};
FixedMemoryStream stream{readonly_bytes};
ReadonlyBytes give_them_back{stream.bytes()};
    // -> Panics on VERIFY(m_writing_enabled);
    // but this is fine:
auto bytes = static_cast<FixedMemoryStream const&>(*stream).bytes()

If we need to be explicit about it, let's rename the overload instead of
adding that `static_cast`.

**LibGfx/PNG: Don't keep a reference to input data for the render step**

Animated image decoded status consists of more information than just a
decoded bitmap. These data needs however to be decoupled from the
original, encoded, data.

**LibGfx/PNG: Don't store the ICC profile as a reference to input data**

The copy can seem unfortunate, but there is no choice in case of a
streamed input.

**LibGfx/PNG: Use `AK::Stream` instead of a custom bit reader**

It allows us to get rid of this dusty `Streamer` and bring us closer to
streamable image decoding!

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/PNGLoader.cpp [hunk 1]
```diff
@@ -128,7 +128,7 @@ enum RenderingIntent {
 };
 
 struct AnimationFrame {
-    fcTL_Chunk const& fcTL;
+    fcTL_Chunk fcTL;
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/PNGLoader.cpp:133]: This looks weird (I'm reading comments in order). Why is this needed?

File: Userland/Libraries/LibGfx/ImageFormats/PNGLoader.cpp [hunk 2]
```diff
@@ -1080,7 +1081,10 @@ static ErrorOr<void> process_iCCP(ReadonlyBytes data, PNGLoadingContext& context
     if (compression_method != 0)
         return Error::from_string_literal("Unsupported compression method in the iCCP chunk");
 
-    context.embedded_icc_profile = EmbeddedICCProfile { { data.data(), profile_name_length }, data.slice(profile_name_length + 2) };
+    auto profile_name = TRY(String::from_utf8({ data.data(), profile_name_length }));
+    auto profile_data = TRY(ByteBuffer::copy(data.slice(profile_name_length + 2)));
+
+    context.embedded_icc_profile = EmbeddedICCProfile { move(profile_name), move(profile_data) };
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/PNGLoader.cpp:1019]: I agree, this seems pretty unfortunate.

**Comment by LucasChollet** [comment]
I've formulated it this way because porting to a Stream make it appears, but there is IMO a fundamental issue here.

Decoders are given `Bytes`, not an owning structure, so they should read them, generate a bitmap and when they give the control back, the caller can do whatever it wants with the input data. Why would you have a copy of both the encoded and decoded image in memory if you just need to display it.

In its current form, the PNG loader relies on the caller to keep the data around, as both the `AnimationFrame` data and the ICC profile are just references to the input stream.

FWIW, `fcTL_Chunk` has a size of 26 bytes, so that's not a big deal, and we are already copying the ICC profile in the JPEG decoder.

> That way you can convert to a stream without having to introduce copies and whatnot.

The only unfortunate copies that I see is that each chunk is copied in `process_chunk`, but that's only because all `process_XXXX` aren't converted yet. I should have added a `FIXME` there, my bad.
I would have considered them as terrible, but a quick profile shows that all these copies makes `process_chunk` goes from 0.02 to 0.05%. They are probably the least of our concerns.

**Comment by nico** [comment]
> Decoders are given Bytes, not an owning structure, so they should read them, generate a bitmap and when they give the control back, the caller can do whatever it wants with the input data. Why would you have a copy of both the encoded and decoded image in memory if you just need to display it.

For the bitmap, sure. But metadata like ICC profile, EXIF data, IPTC, XMP, THumbnails etc is fairly large and likely not used by most clients. So having a clear comment along the lines of "this data is a reference to the input data; make a copy if you need it" in the decoder interface and returning spans might be the better tradeoff there imo.

> FWIW, fcTL_Chunk has a size of 26 bytes, so that's not a big deal, and we are already copying the ICC profile in the JPEG decoder.

Yes, I agree that this change isn't a big deal, it just needs some more details in the commit message. But given the above, it might not be needed anyways.
